lib/signer.ex

```diff
@@ -134,15 +123,45 @@ defmodule AntikytheraAws.Signer do
       |> Enum.into("", fn {key, val} -> "#{key}:#{trimall(val)}\n" end) # Should end with trailing newline
     end
 
-    defunpt signed_headers_string(downcased_headers :: v[Headers.t]) :: String.t do
-      Map.keys(downcased_headers)
-      |> Enum.sort()
-      |> Enum.join(";")
+    defunp compute_signature(aki                :: v[String.t],
+                             sak                :: v[String.t],
+                             amz_date           :: v[AmzDate.t],
+                             region             :: v[String.t],
+                             service            :: v[String.t],
+                             canonical_request  :: v[String.t],
+                             signed_headers_str :: v[String.t]) :: String.t do
+      skey  = signing_key(sak, amz_date, region, service)
+      scope = credential_scope(amz_date, region, service)
+      sts   = string_to_sign(amz_date, scope, canonical_request)
+      sign  = hex_hmac_sha256(skey, sts)
+      "#{@sign_algorithm} Credential=#{aki}/#{scope}, SignedHeaders=#{signed_headers_str}, Signature=#{sign}"
+    end
+
+    defunpt signing_key(sak      :: v[String.t],
+                        amz_date :: v[AmzDate.t],
+                        region   :: v[String.t],
+                        service  :: v[String.t]) :: String.t do
+      short_date = String.slice(amz_date, 0..7)
+      @key_prefix <> sak
+      |> hmac_sha256(short_date)
+      |> hmac_sha256(region)
+      |> hmac_sha256(service)
+      |> hmac_sha256(@termination_string)
+    end
+
+    defunp credential_scope(amz_date :: v[AmzDate.t], region :: v[String.t], service :: v[String.t]) :: String.t do
+      [String.slice(amz_date, 0..7), region, service, @termination_string] |> Enum.join("/")
```

Comment:
(Concerning the original code)
Staring at the code around here, bodies of `signing_key/4` and `credential_scope/4` should be somehow combined.
```ex
signing_key_elements = [String.slice(amz_date, 0..7), region, service, @termination_string]
credential_scope = Enum.join(l, "/")
signing_key = Enum.reduce(l, @key_prefix <> sak, fn(s, k) -> hmac_sha256(k, s) end)
```
You could reverse the argument order of `hmac_sha256/2`.

Reply:
fixed in ccdf04e 
I got lost as to whether to put the processing inside `compute_signature/7` or divide into function, finally I decided to divide into function.